maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java

```diff
@@ -847,12 +847,31 @@ private DefaultScanResult scanDependencies()
         {
             try
             {
+                DefaultScanResult scanResult = new DefaultScanResult( Collections.EMPTY_LIST );
+                
+                List<String> dependenciesToScan = new ArrayList();
+                Collections.addAll( dependenciesToScan, getDependenciesToScan() );
+                TestListResolver includedAndExcludedTests = getIncludedAndExcludedTests();
+
+                for ( MavenProject mavenProject : session.getSortedProjects() )
+                {
+                    String groupArtifactId = mavenProject.getGroupId() + ":" + mavenProject.getArtifactId();
+                    if ( dependenciesToScan.removeAll( Collections.singleton( groupArtifactId ) ) )
+                    {
+                        File outputDirectoryFile = new File( mavenProject.getBuild().getOutputDirectory() );
+                        DirectoryScanner scanner =
+                                new DirectoryScanner( outputDirectoryFile, includedAndExcludedTests );
+                        scanResult = scanResult.append( scanner.scan() );
+                    }
+                }
+
                 // @TODO noinspection unchecked, check MavenProject 3.x for Generics in surefire:3.0
                 @SuppressWarnings( "unchecked" )
-                List<File> dependenciesToScan =
-                    DependencyScanner.filter( project.getTestArtifacts(), Arrays.asList( getDependenciesToScan() ) );
-                DependencyScanner scanner = new DependencyScanner( dependenciesToScan, getIncludedAndExcludedTests() );
-                return scanner.scan();
+                List<File> dependenciesToScanFile =
+                        DependencyScanner.filter( project.getTestArtifacts(), dependenciesToScan );
```

Comment:
Here some artifacts may have classifier `tests`. This means your loop may appear below test artifacts and only those project artifacts (gid:aid:*) should be excluded which appear in `getTestArtifacts()` and scans appended creating one aggregated scan result.

Reply:
Locally staged updates to reflect your generics comments. But I don't follow what you're asking for here.

[Per the documentation](https://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#dependenciesToScan), the dependenciesToScan property doesn't support classifiers - just group ID and artifact ID. So I didn't add any logic around classifier resolution. Is that what you were expecting?